ext/mbstring/mbstring.c

```diff
@@ -2282,6 +2289,256 @@ PHP_FUNCTION(mb_output_handler)
+				}
+			} else{
+				array_init_size(return_value, chunks + 1);
+				for(size_t i = 0; i < chunks; p += chunk_len, ++i){ /* split cycle except last chunk */
+					add_next_index_stringl(return_value, p, chunk_len);
+				}
+				add_next_index_stringl(return_value, p, last - p); /* last chunk */
+			}
+		}
+	} else if (mbfl_encoding->flag & (MBFL_ENCTYPE_WCS4BE | MBFL_ENCTYPE_WCS4LE )) { /* 4 bytes */
+		mb_len = string.len>>2;
+		size_t chunk_len = split_length<<2; /* split length in bytes */
+		if(chunk_len >= string.len){
+			chunk_len = string.len; /* chunk length can't be bigger then string */
+			chunks = 1;
+			array_init_size(return_value, chunks);
+			add_next_index_stringl(return_value, p, chunk_len);
+		} else{
+			chunks = mb_len / (size_t)split_length;
+			if(string.len % chunk_len == 0){
+				array_init_size(return_value, chunks);
+				for(size_t i = 0; i < chunks; p += chunk_len, ++i){
+					add_next_index_stringl(return_value, p, chunk_len);
+				}
+			} else{
+				array_init_size(return_value, chunks + 1);
+				for(size_t i = 0; i < chunks; p += chunk_len, ++i){ /* split cycle except last chunk */
+					add_next_index_stringl(return_value, p, chunk_len);
+				}
+				add_next_index_stringl(return_value, p, last - p); /* last chunk */
+			}
+		}
+	/*
+	 * +----------------------------------------------------------------------+
+	 * | second scenario: "variable width encodings with length table"        |
+	 * +----------------------------------------------------------------------+
+	 */
+	} else if (mbfl_encoding->mblen_table != NULL) {
+		char unsigned const *mbtab = mbfl_encoding->mblen_table;
+		array_init_size(return_value, string.len / split_length); /* init array with maximum possible size */
+
+		while (p < last) { /* split cycle work until the cursor has reached the last byte */
+			char const *chunk_p = p; /* chunk first byte pointer */
+			size_t chunk_length = 0; /* chunk length in bytes */
+			for(uint32_t char_count = 0; char_count < split_length; ++char_count) {
+				char unsigned const m = mbtab[*p]; /* single character length table */
+				chunk_length += m;
+				p += m;
+				++mb_len;
+			}
+			add_next_index_stringl(return_value, chunk_p, chunk_length);
+		}
+	/*
+	 * +----------------------------------------------------------------------+
+	 * | third scenario: "else multibyte encodings without length table"      |
+	 * +----------------------------------------------------------------------+
+	 */
+	} else {
+		mbfl_convert_filter *filter;
+		/* count mb_len */
+		filter = mbfl_convert_filter_new(
+			string.encoding,
+			&mbfl_encoding_wchar,
+			filter_count_output,
+			0,
+			&mb_len);
+		if (filter == NULL){
+			mbfl_convert_filter_delete(filter);
+			RETURN_FALSE; /* something wrong with the filter */
+		}
+		while (p < last) { /* cycle each byte with callback function */
+			(*filter->filter_function)(*p++, filter);
+		}
+		mbfl_convert_filter_delete(filter);
+		/* count mb_len end */
```

Comment:
Why not just call `mbfl_strlen()` for this?

Reply:
For no particular reason.

Reply:
I remember. There is a possible bug in at least 2 mbfl functions including mbfl_strlen. Thing is utf-16 encoding constant missing in condition for 2-bytes encodings.

Reply:
It is NOT a bug. The number of codepoints in UTF-16 string is not equal to numberOfBytes/2. 
Some characters might be encoded as two 16-bit units (surrogate pairs).